{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57a5e96e",
   "metadata": {},
   "source": [
    "# Genes\n",
    "The following notebook retrieves all the genes involved in the reactions from the **\"Rxns\" Sheet** in Google Sheet file. Then, information regarding the genes is retrieved from different databases and a **\"Genes\" Sheet** is generated and updated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08228a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from google_sheet import GoogleSheet\n",
    "from utils import get_gene_info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac38d2d7",
   "metadata": {},
   "source": [
    "### 1. Generate \"rxns\" and \"genes\" datasets\n",
    "The \"rxns\" dataset contains all the reactions with their GPR associations. The \"genes\" dataset contains all the information of the genes/GPR involved in our reconstruction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19d8026b",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEY_FILE_PATH = 'credentials.json'\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sheet_rxns = 'Rxns'\n",
    "sheet_genes = 'Genes'\n",
    "\n",
    "rxns = sheet.read_google_sheet(sheet_rxns)\n",
    "genes = sheet.read_google_sheet(sheet_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de13c811",
   "metadata": {},
   "source": [
    "### 2. Create the a \"gene_list\" with all the genes involved in our reconstruction\n",
    "We extract the gene IDs from the GPR annotations in the **\"rxns\" df**. This information is only retrieved from the **GPR_final** column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0f61d39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3045"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generation of gene_list from all the genes in the \"Whole Cell Network Reconstruction in CHO Cells\" dataset\n",
    "import re\n",
    "\n",
    "gene_list = []\n",
    "human_gene_list = []\n",
    "for index, row in rxns.iterrows():\n",
    "    if row['GPR_final'] != '':\n",
    "        gprs = str(row['GPR_final'])\n",
    "        gpr = re.findall(r'\\b[hH]?\\d+\\b', gprs)\n",
    "        for g in gpr:\n",
    "            if g.lower().startswith('h'):  # we convert to lower case to catch both 'h' and 'H'\n",
    "                human_gene_list.append(int(g[1:]))  # remove 'h' and convert to int before appending\n",
    "            else:\n",
    "                gene_list.append(int(g))  # convert to int before appending\n",
    "\n",
    "gene_list = list(set(gene_list))\n",
    "len(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd3c0d42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build a dictionary with human gene IDs as keys and CHO gene IDs as values\n",
    "orthologs = pd.read_csv('../Data/Orthologs/cho2human_mapping.tsv', sep='\\t')\n",
    "human_to_cho_dict = dict(zip(orthologs['HUMAN_ID'], orthologs['CHO_ID']))\n",
    "\n",
    "def replace_human_ids_with_cho_ids(s):\n",
    "    replaced_dict = {}\n",
    "    if isinstance(s, str):\n",
    "        genes = re.findall(r'\\b[hH]?\\d+\\b', s)\n",
    "        for gene in genes:\n",
    "            if gene.lower().startswith('h'):  # if gene ID starts with 'h' or 'H'\n",
    "                human_id = int(gene[1:])  # remove 'h' and convert to int\n",
    "                if human_id in human_to_cho_dict:  # if human gene ID is in the dictionary\n",
    "                    cho_id = human_to_cho_dict[human_id]  # get corresponding CHO gene ID\n",
    "                    s = s.replace(gene, str(cho_id))  # replace human gene ID with CHO gene ID in the string\n",
    "                    replaced_dict[gene] = cho_id\n",
    "    return s, replaced_dict\n",
    "\n",
    "# Initialize a dictionary to store replacements\n",
    "replacements = {}\n",
    "\n",
    "# Apply the function and update the replacements dictionary\n",
    "for i in range(len(rxns)):\n",
    "    rxns.at[i, 'GPR_final'], replacements_dict = replace_human_ids_with_cho_ids(rxns.at[i, 'GPR_final'])\n",
    "    replacements.update(replacements_dict)\n",
    "\n",
    "# Now, 'replacements' is a dictionary where keys are the original gene IDs and values are the replaced CHO gene IDs\n",
    "replacements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cc309360",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No changes to the original Rxns Sheet\n"
     ]
    }
   ],
   "source": [
    "######################################\n",
    "#### ---------------------------- ####\n",
    "#### ---- Update Rxns Sheet ----- ####\n",
    "#### ---------------------------- ####\n",
    "######################################\n",
    "if replacements:\n",
    "    sheet.update_google_sheet(sheet_rxns, rxns)\n",
    "    print(\"Google Sheet updated.\")\n",
    "elif not replacements:\n",
    "    print(\"No changes to the original Rxns Sheet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed00e36d",
   "metadata": {},
   "source": [
    "### 3. Add the genes from the \"gene_list\" to the genes df\n",
    "Using a list of all the genes included in the dataset we can retrieve information from the NIH database regarding Gene Symbol, Gene Name, Gene Ensembl ID, and mRNA ID and protein ID using the function get_gene_info()."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "236d1799",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "88dcf69b4fa343dca5f8865c837e50fd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3045 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100769795\n",
      "100761603\n",
      "100770004\n",
      "100770100\n",
      "100761930\n",
      "100753739\n",
      "Gene 100753739 does not have PICR Ensembl ID\n",
      "100770138\n",
      "100753824\n",
      "100770274\n",
      "100753962\n",
      "103162531\n",
      "Gene 103162531 does not have PICR Ensembl ID\n",
      "100754136\n",
      "100762374\n",
      "100754196\n",
      "100754357\n",
      "100762553\n",
      "100754451\n",
      "100770876\n",
      "100770922\n",
      "Gene 100770922 does not have PICR Ensembl ID\n",
      "100771070\n",
      "100754734\n",
      "100771270\n",
      "103163344\n",
      "Gene 103163344 does not have PICR Ensembl ID\n",
      "100763215\n",
      "100763252\n",
      "100771461\n",
      "100771582\n",
      "100763403\n",
      "103163675\n",
      "Gene 103163675 does not have PICR Ensembl ID\n",
      "100763430\n",
      "100763474\n",
      "100755288\n",
      "100771780\n",
      "100763595\n",
      "100755641\n",
      "Gene 100755641 does not have PICR Ensembl ID\n",
      "100755739\n",
      "100755837\n",
      "100755863\n",
      "100764057\n",
      "Gene 100764057 does not have PICR Ensembl ID\n",
      "100764142\n",
      "Gene 100764142 does not have PICR Ensembl ID\n",
      "Gene 100764142 does not have mRNA id and protein id\n",
      "100772358\n",
      "100755998\n",
      "100764216\n",
      "100764593\n",
      "100764809\n",
      "Gene 100764809 does not have PICR Ensembl ID\n",
      "100773010\n",
      "Gene 100773010 does not have PICR Ensembl ID\n",
      "100764848\n",
      "100764916\n",
      "100765169\n",
      "100773417\n",
      "100773511\n",
      "Gene 100773511 does not have PICR Ensembl ID\n",
      "100757160\n",
      "100773599\n",
      "100757223\n",
      "100773691\n",
      "100773712\n",
      "100757368\n",
      "100765719\n",
      "Gene 100765719 does not have PICR Ensembl ID\n",
      "100757543\n",
      "100774056\n",
      "100774146\n",
      "Gene 100774146 does not have mRNA id and protein id\n",
      "100757792\n",
      "100766062\n",
      "100766065\n",
      "30100754813\n",
      "Gene 30100754813 does not have PICR Ensembl ID\n",
      "Gene (30100754813, 'ycf15') is a Vitis cinerea Gene\n",
      "100774394\n",
      "100774547\n",
      "100758229\n",
      "Gene 100758229 does not have PICR Ensembl ID\n",
      "100758361\n",
      "100774757\n",
      "100774844\n",
      "100766720\n",
      "Gene 100766720 does not have PICR Ensembl ID\n",
      "100766772\n",
      "100767001\n",
      "100758827\n",
      "Gene 100758827 does not have PICR Ensembl ID\n",
      "100750691\n",
      "100767149\n",
      "100759190\n",
      "100767485\n",
      "100751123\n",
      "100759338\n",
      "100751175\n",
      "Gene 100751175 does not have PICR Ensembl ID\n",
      "103159723\n",
      "Gene 103159723 does not have PICR Ensembl ID\n",
      "100751325\n",
      "100751347\n",
      "100751558\n",
      "Gene 100751558 does not have PICR Ensembl ID\n",
      "100759761\n",
      "100751755\n",
      "100751960\n",
      "100768429\n",
      "100768553\n",
      "103160652\n",
      "Gene 103160652 does not have PICR Ensembl ID\n",
      "100768669\n",
      "100752432\n",
      "100769016\n",
      "100769171\n",
      "100761077\n",
      "100769277\n",
      "100769454\n",
      "Gene 100769454 does not have PICR Ensembl ID\n",
      "100761300\n",
      "Gene 100761300 does not have PICR Ensembl ID\n",
      "100761353\n",
      "100753164\n",
      "100761387\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "from urllib.error import HTTPError\n",
    "from Bio.Entrez.Parser import ValidationError\n",
    "\n",
    "max_retries = 5  # Set the maximum number of retries\n",
    "\n",
    "for g in tqdm(gene_list):\n",
    "    if str(g) not in list(genes['Gene Entrez ID']):\n",
    "        retry = True\n",
    "        retries = 0\n",
    "        while retry:\n",
    "            try:\n",
    "                print(g)\n",
    "                organism, gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "                if organism == 'Cricetulus griseus':\n",
    "                    new_row_data = {'Gene Entrez ID': g, 'Gene Symbol': gene_symbol, 'Gene Name': gene_name, 'Gene Description': gene_description,\n",
    "                                    'PICR Ensembl ID': picr_ensembl_id, 'CHOK1GS Ensembl ID': chok1gs_ensembl_id, 'Transcript ID': mRNA_ncbi_id,\n",
    "                                    'Protein ID': protein_ncbi_id, 'GO Terms': go_terms}\n",
    "                    new_row_df = pd.DataFrame(new_row_data, index=[len(genes)])\n",
    "                    genes = pd.concat([genes, new_row_df])\n",
    "                    retry = False\n",
    "                elif organism != 'Cricetulus griseus':\n",
    "                    print(f'Gene {g,gene_symbol} is a {organism} Gene')\n",
    "                    break\n",
    "            except ValidationError:\n",
    "                print(f'Gene {g} not found')\n",
    "                retry = False\n",
    "                continue\n",
    "            except HTTPError:\n",
    "                retries += 1\n",
    "                if retries > max_retries:\n",
    "                    print(f'HTTP Error, reached maximum retries ({max_retries}) for gene {g}')\n",
    "                    break\n",
    "                else:\n",
    "                    print(f'HTTP Error, retrying with gene {g}')\n",
    "                    time.sleep(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3c320189",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Entrez ID</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Description</th>\n",
       "      <th>PICR Ensembl ID</th>\n",
       "      <th>CHOK1GS Ensembl ID</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>Protein ID</th>\n",
       "      <th>GO Terms</th>\n",
       "      <th>Uniprot ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100774298</td>\n",
       "      <td>Slco1b3</td>\n",
       "      <td>solute carrier organic anion transporter famil...</td>\n",
       "      <td>This gene encodes a liver-specific member of t...</td>\n",
       "      <td>ENSCGRG00015012813</td>\n",
       "      <td>ENSCGRG00001021568</td>\n",
       "      <td>XM_035456357</td>\n",
       "      <td>XP_035312248</td>\n",
       "      <td>['GO:0055085', 'GO:0022857', 'GO:0005515', 'GO...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100754260</td>\n",
       "      <td>Hsd3b7</td>\n",
       "      <td>hydroxy-delta-5-steroid dehydrogenase, 3 beta-...</td>\n",
       "      <td>This gene encodes an enzyme which is involved ...</td>\n",
       "      <td>ENSCGRG00015024104</td>\n",
       "      <td>ENSCGRG00001010375</td>\n",
       "      <td>XM_003510980</td>\n",
       "      <td>XP_003511028</td>\n",
       "      <td>['GO:0016020', 'GO:0016491', 'GO:0016616', 'GO...</td>\n",
       "      <td>G3I7C5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100769024</td>\n",
       "      <td>LOC100769024</td>\n",
       "      <td>aromatase</td>\n",
       "      <td>This gene encodes a member of the cytochrome P...</td>\n",
       "      <td>ENSCGRG00015026385</td>\n",
       "      <td>ENSCGRG00001015094</td>\n",
       "      <td>XM_003508783</td>\n",
       "      <td>XP_003508831</td>\n",
       "      <td>['GO:0016020', 'GO:0046872', 'GO:0016491', 'GO...</td>\n",
       "      <td>A0A061IC63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100756356</td>\n",
       "      <td>Atp2a3</td>\n",
       "      <td>ATPase sarcoplasmic/endoplasmic reticulum Ca2+...</td>\n",
       "      <td>This gene encodes one of the SERCA Ca(2+)-ATPa...</td>\n",
       "      <td>ENSCGRG00015000763</td>\n",
       "      <td>ENSCGRG00001017225</td>\n",
       "      <td>XM_035458322</td>\n",
       "      <td>XP_035314213</td>\n",
       "      <td>['GO:0005783', 'GO:0016529', 'GO:0044325', 'GO...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100774660</td>\n",
       "      <td>Pgls</td>\n",
       "      <td>6-phosphogluconolactonase</td>\n",
       "      <td>Enables 6-phosphogluconolactonase activity. In...</td>\n",
       "      <td>ENSCGRG00015004513</td>\n",
       "      <td>ENSCGRG00001024276</td>\n",
       "      <td>XM_027428449</td>\n",
       "      <td>XP_027284250</td>\n",
       "      <td>['GO:0016787', 'GO:0005975', 'GO:0017057', 'GO...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3177</th>\n",
       "      <td>100769454</td>\n",
       "      <td>LOC100769454</td>\n",
       "      <td>phosphatidylinositol N-acetylglucosaminyltrans...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>XM_003515280</td>\n",
       "      <td>XP_003515328</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3178</th>\n",
       "      <td>100761300</td>\n",
       "      <td>LOC100761300</td>\n",
       "      <td>histamine N-methyltransferase</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>XM_003509092</td>\n",
       "      <td>XP_003509140</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3179</th>\n",
       "      <td>100761353</td>\n",
       "      <td>Ptprg</td>\n",
       "      <td>protein tyrosine phosphatase receptor type G</td>\n",
       "      <td></td>\n",
       "      <td>ENSCGRG00015014841</td>\n",
       "      <td></td>\n",
       "      <td>XM_035454322</td>\n",
       "      <td>XP_035310213</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3180</th>\n",
       "      <td>100753164</td>\n",
       "      <td>LOC100753164</td>\n",
       "      <td>bile salt sulfotransferase 1-like</td>\n",
       "      <td></td>\n",
       "      <td>ENSCGRG00015042115</td>\n",
       "      <td></td>\n",
       "      <td>XM_035437148</td>\n",
       "      <td>XP_035293039</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3181</th>\n",
       "      <td>100761387</td>\n",
       "      <td>Serpina1</td>\n",
       "      <td>serpin family A member 1</td>\n",
       "      <td></td>\n",
       "      <td>ENSCGRG00015015273</td>\n",
       "      <td></td>\n",
       "      <td>XM_035460417</td>\n",
       "      <td>XP_035316308</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3182 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene Entrez ID   Gene Symbol  \\\n",
       "0         100774298       Slco1b3   \n",
       "1         100754260        Hsd3b7   \n",
       "2         100769024  LOC100769024   \n",
       "3         100756356        Atp2a3   \n",
       "4         100774660          Pgls   \n",
       "...             ...           ...   \n",
       "3177      100769454  LOC100769454   \n",
       "3178      100761300  LOC100761300   \n",
       "3179      100761353         Ptprg   \n",
       "3180      100753164  LOC100753164   \n",
       "3181      100761387      Serpina1   \n",
       "\n",
       "                                              Gene Name  \\\n",
       "0     solute carrier organic anion transporter famil...   \n",
       "1     hydroxy-delta-5-steroid dehydrogenase, 3 beta-...   \n",
       "2                                             aromatase   \n",
       "3     ATPase sarcoplasmic/endoplasmic reticulum Ca2+...   \n",
       "4                             6-phosphogluconolactonase   \n",
       "...                                                 ...   \n",
       "3177  phosphatidylinositol N-acetylglucosaminyltrans...   \n",
       "3178                      histamine N-methyltransferase   \n",
       "3179       protein tyrosine phosphatase receptor type G   \n",
       "3180                  bile salt sulfotransferase 1-like   \n",
       "3181                           serpin family A member 1   \n",
       "\n",
       "                                       Gene Description     PICR Ensembl ID  \\\n",
       "0     This gene encodes a liver-specific member of t...  ENSCGRG00015012813   \n",
       "1     This gene encodes an enzyme which is involved ...  ENSCGRG00015024104   \n",
       "2     This gene encodes a member of the cytochrome P...  ENSCGRG00015026385   \n",
       "3     This gene encodes one of the SERCA Ca(2+)-ATPa...  ENSCGRG00015000763   \n",
       "4     Enables 6-phosphogluconolactonase activity. In...  ENSCGRG00015004513   \n",
       "...                                                 ...                 ...   \n",
       "3177                                                                          \n",
       "3178                                                                          \n",
       "3179                                                     ENSCGRG00015014841   \n",
       "3180                                                     ENSCGRG00015042115   \n",
       "3181                                                     ENSCGRG00015015273   \n",
       "\n",
       "      CHOK1GS Ensembl ID Transcript ID    Protein ID  \\\n",
       "0     ENSCGRG00001021568  XM_035456357  XP_035312248   \n",
       "1     ENSCGRG00001010375  XM_003510980  XP_003511028   \n",
       "2     ENSCGRG00001015094  XM_003508783  XP_003508831   \n",
       "3     ENSCGRG00001017225  XM_035458322  XP_035314213   \n",
       "4     ENSCGRG00001024276  XM_027428449  XP_027284250   \n",
       "...                  ...           ...           ...   \n",
       "3177                      XM_003515280  XP_003515328   \n",
       "3178                      XM_003509092  XP_003509140   \n",
       "3179                      XM_035454322  XP_035310213   \n",
       "3180                      XM_035437148  XP_035293039   \n",
       "3181                      XM_035460417  XP_035316308   \n",
       "\n",
       "                                               GO Terms  Uniprot ID  \n",
       "0     ['GO:0055085', 'GO:0022857', 'GO:0005515', 'GO...              \n",
       "1     ['GO:0016020', 'GO:0016491', 'GO:0016616', 'GO...      G3I7C5  \n",
       "2     ['GO:0016020', 'GO:0046872', 'GO:0016491', 'GO...  A0A061IC63  \n",
       "3     ['GO:0005783', 'GO:0016529', 'GO:0044325', 'GO...              \n",
       "4     ['GO:0016787', 'GO:0005975', 'GO:0017057', 'GO...              \n",
       "...                                                 ...         ...  \n",
       "3177                                                            NaN  \n",
       "3178                                                            NaN  \n",
       "3179                                                            NaN  \n",
       "3180                                                            NaN  \n",
       "3181                                                            NaN  \n",
       "\n",
       "[3182 rows x 10 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2218925",
   "metadata": {},
   "source": [
    "### 4. Eliminate unwanted genes\n",
    "We iterate over the entire dataset to spot **Human genes** in order to eliminate them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b810d3c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a copy of the DataFrame to avoid changing the original while iterating\n",
    "genes_copy = genes.copy()\n",
    "\n",
    "for index, row in tqdm(genes.iterrows(), total=genes.shape[0]):\n",
    "    g = row['Gene Entrez ID']\n",
    "    retry = True\n",
    "    retries = 0\n",
    "    while retry:\n",
    "        try:\n",
    "            organism, gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "            if organism == 'Homo sapiens':\n",
    "                genes_copy = genes_copy.drop(index)\n",
    "                print(f'Gene {g,gene_symbol} erased from dataset')\n",
    "                retry = False\n",
    "            else:\n",
    "                retry = False\n",
    "        except ValidationError:\n",
    "            print(f'Gene {g} not found')\n",
    "            retry = False\n",
    "            continue\n",
    "        except HTTPError:\n",
    "            retries += 1\n",
    "            if retries > max_retries:\n",
    "                print(f'HTTP Error, reached maximum retries ({max_retries}) for gene {g}')\n",
    "                break\n",
    "            else:\n",
    "                print(f'HTTP Error, retrying with gene {g}')\n",
    "                time.sleep(1)\n",
    "\n",
    "# Replace the original DataFrame with the updated one\n",
    "genes = genes_copy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a459ea77",
   "metadata": {},
   "outputs": [],
   "source": [
    "sheet.update_google_sheet(sheet_genes, genes)\n",
    "print(\"Google Sheet updated.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4536998b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbc53f26",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5abc29b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f3b506a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in tqdm(gene_list):\n",
    "    if g not in list(genes['Gene Entrez ID']):\n",
    "        try:\n",
    "            gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "            new_row_data = {'Gene Entrez ID': g, 'Gene Symbol': gene_symbol, 'Gene Name': gene_name, 'Gene Description': gene_description,\n",
    "                            'PICR Ensembl ID': picr_ensembl_id, 'CHOK1GS Ensembl ID': chok1gs_ensembl_id, 'Transcript ID': mRNA_ncbi_id,\n",
    "                            'Protein ID': protein_ncbi_id, 'GO Terms': go_terms}\n",
    "            new_row_df = pd.DataFrame(new_row_data, index=[len(genes)])\n",
    "            genes = pd.concat([genes, new_row_df])\n",
    "        except KeyError:\n",
    "            print(f'Gene {g} not found')\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dfcbfb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch information from the NIH database\n",
    "\n",
    "\n",
    "# Complete null or blank information in the already generated \"Genes Sheet\" dataset\n",
    "for i,row in genes.iterrows():\n",
    "    if row['Gene Entrez ID'] == '':\n",
    "        for g in gene_list:\n",
    "            gene_sheet_list = [str(x) for x in genes['Gene Entrez ID']]\n",
    "            if g not in gene_sheet_list:\n",
    "                gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "                row['Gene Entrez ID'] = g\n",
    "                row['Gene Symbol'] = gene_symbol\n",
    "                row['Gene Name'] = gene_name\n",
    "                row['Gene Description'] = gene_description\n",
    "                row['PICR Ensembl ID'] = picr_ensembl_id\n",
    "                row['CHOK1GS Ensembl ID'] = chok1gs_ensembl_id\n",
    "                row['Transcript ID'] = mRNA_ncbi_id\n",
    "                row['Protein ID'] = protein_ncbi_id\n",
    "                row['GO Terms'] = go_terms\n",
    "                break          \n",
    "    elif row['Gene Entrez ID'] != '' and (row['Gene Symbol'] == '' or row['Gene Name'] == '' or row['PICR Ensembl ID'] == '' or row['Transcript ID'] == '' or row['Protein ID'] == ''):\n",
    "        row['Gene Symbol'] = gene_symbol\n",
    "        row['Gene Name'] = gene_name\n",
    "        row['Gene Description'] = gene_description\n",
    "        row['PICR Ensembl ID'] = picr_ensembl_id\n",
    "        row['CHOK1GS Ensembl ID'] = chok1gs_ensembl_id\n",
    "        row['Transcript ID'] = mRNA_ncbi_id\n",
    "        row['Protein ID'] = protein_ncbi_id\n",
    "        row['GO Terms'] = go_terms\n",
    "'''\n",
    "# Add genes from the gene_list that are not yet in the \"Genes Sheet\" dataset\n",
    "for g in gene_list:\n",
    "    # the first try/except is to avoid overwritting data in case there already some info in the dataset\n",
    "    try:\n",
    "        gene_sheet_list = [str(x) for x in df['Gene Entrez ID']]\n",
    "        id = max(df['Index']) + 2\n",
    "    except:\n",
    "        gene_sheet_list = []\n",
    "        id = 2\n",
    "    if gene not in gene_sheet_list:\n",
    "        try:\n",
    "            gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(gene)\n",
    "            print(id-1)\n",
    "            genes_sheet.update_cell(id,1,id-1)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,2,gene)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,3,gene_symbol)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,4,gene_name)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,5,gene_description)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,6,picr_ensembl_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,7,chok1gs_ensembl_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,8,mRNA_ncbi_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,9,protein_ncbi_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,10,go_terms)\n",
    "        except:\n",
    "            print('Google API quota exceeded')\n",
    "            time.sleep(5)\n",
    "            continue\n",
    "            \n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f702b0d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73daf8e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_gene_info('100750772')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30b59d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = 'account1@theta-ocean-377718.iam.gserviceaccount.com'\n",
    "handle = Entrez.efetch(db='gene', id='3979190', retmode='xml')\n",
    "record = Entrez.read(handle)[0]\n",
    "\n",
    "#gene_name = record['Entrezgene_gene']['Gene-ref']['Gene-ref_desc']\n",
    "#gene_symbol = record['Entrezgene_gene']['Gene-ref']['Gene-ref_locus']\n",
    "\n",
    "human_handle = Entrez.efetch(db='gene', id='3979190', retmode='xml')\n",
    "human_record = Entrez.read(human_handle)[0]\n",
    "if 'Entrezgene_comments' in human_record and 'Gene-commentary_comment' in human_record['Entrezgene_comments'][0]:\n",
    "    human_gene_description = human_record['Entrezgene_comments'][0]['Gene-commentary_comment'][0]['String']\n",
    "elif 'Entrezgene_summary' in human_record:\n",
    "    human_gene_description = human_record['Entrezgene_summary']\n",
    "else:\n",
    "    human_gene_description = ''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
