{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57a5e96e",
   "metadata": {},
   "source": [
    "# Genes\n",
    "The following notebook retrieves all the genes involved in the reactions from the **\"Rxns\" Sheet** in Google Sheet file. Then, information regarding the genes is retrieved from different databases and a **\"Genes\" Sheet** is generated and updated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08228a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from google_sheet import GoogleSheet\n",
    "from utils import get_gene_info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac38d2d7",
   "metadata": {},
   "source": [
    "### 1.Generate \"rxns\" and \"genes\" datasets\n",
    "The \"rxns\" dataset contains all the reactions with their GPR associations. The \"genes\" dataset contains all the information of the genes/GPR involved in our reconstruction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19d8026b",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEY_FILE_PATH = 'credentials.json'\n",
    "SPREADSHEET_ID = '1MlBXeHIKw8k8fZyXm-sN__AHTRSunJxar_-bqvukZws'\n",
    "\n",
    "# Initialize the GoogleSheet object\n",
    "sheet = GoogleSheet(SPREADSHEET_ID, KEY_FILE_PATH)\n",
    "\n",
    "# Read data from the Google Sheet\n",
    "sheet_rxns = 'Rxns'\n",
    "sheet_genes = 'Genes'\n",
    "\n",
    "rxns = sheet.read_google_sheet(sheet_rxns)\n",
    "genes = sheet.read_google_sheet(sheet_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20748fad-9a63-4ff6-9297-bc63083e43bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sheet_list = list(genes['Gene Entrez ID'])\n",
    "gene_sheet_list_int = list(map(int, gene_sheet_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de13c811",
   "metadata": {},
   "source": [
    "### 2. Create the a \"gene_list\" with all the genes involved in our reconstruction\n",
    "We extract the gene IDs from the GPR annotations in the **\"rxns\" df**. This information is only retrieved from the **GPR_final** column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0f61d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generation of gene_list from all the genes in the \"Whole Cell Network Reconstruction in CHO Cells\" dataset\n",
    "import re\n",
    "\n",
    "gene_list = []\n",
    "human_gene_list = []\n",
    "for index, row in rxns.iterrows():\n",
    "    if row['GPR_final'] != '':\n",
    "        gprs = str(row['GPR_final'])\n",
    "        gpr = re.findall(r'\\b[hH]?\\d+\\b', gprs)\n",
    "        for g in gpr:\n",
    "            if g.lower().startswith('h'):  # we convert to lower case to catch both 'h' and 'H'\n",
    "                human_gene_list.append(int(g[1:]))  # remove 'h' and convert to int before appending\n",
    "            else:\n",
    "                gene_list.append(int(g))  # convert to int before appending\n",
    "\n",
    "gene_list = list(set(gene_list))\n",
    "len(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0864863-033e-4a9a-a775-2faaa0bfc2a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "set1 = set(gene_sheet_list_int)\n",
    "set2 = set(gene_list)\n",
    "\n",
    "# Find elements unique to each list\n",
    "unique_to_list1 = set1 - set2\n",
    "unique_to_list2 = set2 - set1\n",
    "\n",
    "# Optionally, combine the unique elements\n",
    "unique_elements = unique_to_list1.union(unique_to_list2)\n",
    "\n",
    "# Output the results\n",
    "print(f\"Elements unique to list1: {unique_to_list1}\")\n",
    "print(f\"Elements unique to list2: {unique_to_list2}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c5c853",
   "metadata": {},
   "source": [
    "### 3. Replace remaining Human Gene IDs in our dataset with CHO Gene IDs\n",
    "Using the **cho2human_mapping** file, we replace the Human gene IDs with those of CHO, these IDs were not spotted when we first added the Recon3D reactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd3c0d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a dictionary with human gene IDs as keys and CHO gene IDs as values\n",
    "orthologs = pd.read_csv('../Data/Orthologs/cho2human_mapping.tsv', sep='\\t')\n",
    "human_to_cho_dict = dict(zip(orthologs['HUMAN_ID'], orthologs['CHO_ID']))\n",
    "\n",
    "def replace_human_ids_with_cho_ids(s):\n",
    "    replaced_dict = {}\n",
    "    if isinstance(s, str):\n",
    "        genes = re.findall(r'\\b[hH]?\\d+\\b', s)\n",
    "        for gene in genes:\n",
    "            if gene.lower().startswith('h'):  # if gene ID starts with 'h' or 'H'\n",
    "                human_id = int(gene[1:])  # remove 'h' and convert to int\n",
    "                if human_id in human_to_cho_dict:  # if human gene ID is in the dictionary\n",
    "                    cho_id = human_to_cho_dict[human_id]  # get corresponding CHO gene ID\n",
    "                    s = s.replace(gene, str(cho_id))  # replace human gene ID with CHO gene ID in the string\n",
    "                    replaced_dict[gene] = cho_id\n",
    "    return s, replaced_dict\n",
    "\n",
    "# Initialize a dictionary to store replacements\n",
    "replacements = {}\n",
    "\n",
    "# Apply the function and update the replacements dictionary\n",
    "for i in range(len(rxns)):\n",
    "    rxns.at[i, 'GPR_final'], replacements_dict = replace_human_ids_with_cho_ids(rxns.at[i, 'GPR_final'])\n",
    "    replacements.update(replacements_dict)\n",
    "\n",
    "# Now, 'replacements' is a dictionary where keys are the original gene IDs and values are the replaced CHO gene IDs\n",
    "replacements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc309360",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################\n",
    "#### ---------------------------- ####\n",
    "#### ---- Update Rxns Sheet ----- ####\n",
    "#### ---------------------------- ####\n",
    "######################################\n",
    "if replacements:\n",
    "    sheet.update_google_sheet(sheet_rxns, rxns)\n",
    "    print(\"Google Sheet updated.\")\n",
    "elif not replacements:\n",
    "    print(\"No changes to the original Rxns Sheet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed00e36d",
   "metadata": {},
   "source": [
    "### 3. Add the genes from the \"gene_list\" to the genes df\n",
    "Using a list of all the genes included in the dataset we can retrieve information from the NIH database regarding Gene Symbol, Gene Name, Gene Ensembl ID, and mRNA ID and protein ID using the function get_gene_info()."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "236d1799",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "from urllib.error import HTTPError\n",
    "from Bio.Entrez.Parser import ValidationError\n",
    "\n",
    "max_retries = 5  # Set the maximum number of retries\n",
    "c = 0\n",
    "\n",
    "for g in tqdm(gene_list):\n",
    "    if str(g) not in list(genes['Gene Entrez ID']):\n",
    "        retry = True\n",
    "        retries = 0\n",
    "        while retry:\n",
    "            try:\n",
    "                print(g)\n",
    "                organism, gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "                if organism == 'Cricetulus griseus':\n",
    "                    new_row_data = {'Gene Entrez ID': g, 'Gene Symbol': gene_symbol, 'Gene Name': gene_name, 'Gene Description': gene_description,\n",
    "                                    'PICR Ensembl ID': picr_ensembl_id, 'CHOK1GS Ensembl ID': chok1gs_ensembl_id, 'Transcript ID': mRNA_ncbi_id,\n",
    "                                    'Protein ID': protein_ncbi_id, 'GO Terms': go_terms}\n",
    "                    new_row_df = pd.DataFrame(new_row_data, index=[len(genes)])\n",
    "                    genes = pd.concat([genes, new_row_df])\n",
    "                    c+=1\n",
    "                    retry = False\n",
    "                elif organism != 'Cricetulus griseus':\n",
    "                    print(f'Gene {g,gene_symbol} is a {organism} Gene')\n",
    "                    break\n",
    "            except ValidationError:\n",
    "                print(f'Gene {g} not found')\n",
    "                retry = False\n",
    "                continue\n",
    "            except HTTPError:\n",
    "                retries += 1\n",
    "                if retries > max_retries:\n",
    "                    print(f'HTTP Error, reached maximum retries ({max_retries}) for gene {g}')\n",
    "                    break\n",
    "                else:\n",
    "                    print(f'HTTP Error, retrying with gene {g}')\n",
    "                    time.sleep(1)\n",
    "                    \n",
    "print(f'A total of {c} Genes were added to the dataset')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2abe8ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################\n",
    "#### ----------------------------- ####\n",
    "#### ---- Update Genes Sheet ----- ####\n",
    "#### ----------------------------- ####\n",
    "#######################################\n",
    "if c > 0:\n",
    "    sheet.update_google_sheet(sheet_genes, genes)\n",
    "    print(\"Google Sheet updated.\")\n",
    "elif c == 0:\n",
    "    print(\"No new genes added.\")\n",
    "    print(\"No changes to the original Genes Sheet.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2218925",
   "metadata": {},
   "source": [
    "### 4. Eliminate unwanted genes\n",
    "We iterate over the entire dataset to spot **Human genes** in order to eliminate them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d90b73a7-c9ae-4113-b092-3edc2bbe65d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a copy of the DataFrame to avoid changing the original while iterating\n",
    "erased = 0\n",
    "genes_copy = genes.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b810d3c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in tqdm(genes.iterrows(), total=genes.shape[0]):\n",
    "    g = row['Gene Entrez ID']\n",
    "    retry = True\n",
    "    retries = 0\n",
    "    while retry:\n",
    "        try:\n",
    "            organism, gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "            if organism == 'Homo sapiens':\n",
    "                genes_copy = genes_copy.drop(index)\n",
    "                print(f'Gene {g,gene_symbol} erased from dataset')\n",
    "                erased+=1\n",
    "                retry = False\n",
    "            else:\n",
    "                retry = False\n",
    "        except ValidationError:\n",
    "            print(f'Gene {g} not found')\n",
    "            retry = False\n",
    "            continue\n",
    "        except HTTPError:\n",
    "            retries += 1\n",
    "            if retries > max_retries:\n",
    "                print(f'HTTP Error, reached maximum retries ({max_retries}) for gene {g}')\n",
    "                break\n",
    "            else:\n",
    "                print(f'HTTP Error, retrying with gene {g}')\n",
    "                time.sleep(1)\n",
    "\n",
    "# Replace the original DataFrame with the updated one\n",
    "print(f'A total of {erased} Human Genes were identifed from the original dataset')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29b273fb-3d07-4286-9937-0ee4af7136ed",
   "metadata": {},
   "source": [
    "### 5. Eliminate genes that are not in the model\n",
    "Next, we make sure that the genes in the **GPR_final** column of the Reactions Sheet matches those of the **Gene Entrez ID** column of the Genes Sheet. If there are genes in the Genes Sheet that are not in the **GPR_final** column of th Reactions Sheet, we eliminate them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abd8d332-e943-43e9-a764-f1b2f8571066",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for index, row in genes_copy.iterrows():\n",
    "    g = int(row['Gene Entrez ID'])\n",
    "    if g not in gene_list:\n",
    "        genes_copy = genes_copy.drop(index)\n",
    "        erased+=1\n",
    "        print(g)\n",
    "    \n",
    "\n",
    "# Replace the original DataFrame with the updated one\n",
    "print(f'A total of {erased} Genes were erased from the original dataset')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae91d4d5-c76a-46df-b718-62c6bbb2cb34",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = genes_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a459ea77",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################\n",
    "#### ----------------------------- ####\n",
    "#### ---- Update Genes Sheet ----- ####\n",
    "#### ----------------------------- ####\n",
    "#######################################\n",
    "if erased > 0:\n",
    "    sheet.update_google_sheet(sheet_genes, genes)\n",
    "    print(\"Google Sheet updated.\")\n",
    "elif erased == 0:\n",
    "    print(\"No erased genes.\")\n",
    "    print(\"No changes to the original Genes Sheet.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c193455-c5ee-4039-8681-2326b2e5d0b1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### Compare Genes Sheet with Rxns Sheet to make sure we have the same gene IDs in both datasets\n",
    "genes_sheet_list = list(genes['Gene Entrez ID'])\n",
    "genes_sheet_list = [int(item) for item in genes_sheet_list]\n",
    "genes_sheet_list = set(genes_sheet_list)\n",
    "gene_list = set(gene_list)\n",
    "\n",
    "# Find elements in list_a but not in list_b\n",
    "missing_in_genes_sheet = gene_list - genes_sheet_list\n",
    "print(\"Missing in Genes Sheet:\", missing_in_genes_sheet)\n",
    "\n",
    "# Find elements in list_b but not in list_a\n",
    "missing_in_rxns_sheet = genes_sheet_list - gene_list\n",
    "print(\"Missing in Rxns Sheet:\", missing_in_rxns_sheet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ad9748-4b61-4a17-9cad-edeb9d16bf28",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_duplicates(lst):\n",
    "    seen = set()\n",
    "    duplicates = set()\n",
    "    for item in lst:\n",
    "        if item in seen:\n",
    "            duplicates.add(item)\n",
    "        else:\n",
    "            seen.add(item)\n",
    "    return duplicates\n",
    "\n",
    "# Example usage:\n",
    "my_list = list(genes['Gene Entrez ID'])\n",
    "print(find_duplicates(my_list))  # Shows the duplicated items"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1988fa2d-d96b-4856-8e08-851981ffdb6d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06685081-8433-4ea7-8fbf-e7824d264a9b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "167e07ea-2289-4d20-82aa-6883c21196f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6a27047-d265-43d8-9051-1fde6a324030",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdd2f4ce-5507-4c19-b90b-b330d3d6c45a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "265eb92c-0ba9-4ce6-862e-ee5f742d48b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dfcbfb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch information from the NIH database\n",
    "\n",
    "\n",
    "# Complete null or blank information in the already generated \"Genes Sheet\" dataset\n",
    "for i,row in genes.iterrows():\n",
    "    if row['Gene Entrez ID'] == '':\n",
    "        for g in gene_list:\n",
    "            gene_sheet_list = [str(x) for x in genes['Gene Entrez ID']]\n",
    "            if g not in gene_sheet_list:\n",
    "                gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(g)\n",
    "                row['Gene Entrez ID'] = g\n",
    "                row['Gene Symbol'] = gene_symbol\n",
    "                row['Gene Name'] = gene_name\n",
    "                row['Gene Description'] = gene_description\n",
    "                row['PICR Ensembl ID'] = picr_ensembl_id\n",
    "                row['CHOK1GS Ensembl ID'] = chok1gs_ensembl_id\n",
    "                row['Transcript ID'] = mRNA_ncbi_id\n",
    "                row['Protein ID'] = protein_ncbi_id\n",
    "                row['GO Terms'] = go_terms\n",
    "                break          \n",
    "    elif row['Gene Entrez ID'] != '' and (row['Gene Symbol'] == '' or row['Gene Name'] == '' or row['PICR Ensembl ID'] == '' or row['Transcript ID'] == '' or row['Protein ID'] == ''):\n",
    "        row['Gene Symbol'] = gene_symbol\n",
    "        row['Gene Name'] = gene_name\n",
    "        row['Gene Description'] = gene_description\n",
    "        row['PICR Ensembl ID'] = picr_ensembl_id\n",
    "        row['CHOK1GS Ensembl ID'] = chok1gs_ensembl_id\n",
    "        row['Transcript ID'] = mRNA_ncbi_id\n",
    "        row['Protein ID'] = protein_ncbi_id\n",
    "        row['GO Terms'] = go_terms\n",
    "'''\n",
    "# Add genes from the gene_list that are not yet in the \"Genes Sheet\" dataset\n",
    "for g in gene_list:\n",
    "    # the first try/except is to avoid overwritting data in case there already some info in the dataset\n",
    "    try:\n",
    "        gene_sheet_list = [str(x) for x in df['Gene Entrez ID']]\n",
    "        id = max(df['Index']) + 2\n",
    "    except:\n",
    "        gene_sheet_list = []\n",
    "        id = 2\n",
    "    if gene not in gene_sheet_list:\n",
    "        try:\n",
    "            gene_symbol, gene_name, gene_description, picr_ensembl_id, chok1gs_ensembl_id, mRNA_ncbi_id, protein_ncbi_id, go_terms = get_gene_info(gene)\n",
    "            print(id-1)\n",
    "            genes_sheet.update_cell(id,1,id-1)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,2,gene)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,3,gene_symbol)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,4,gene_name)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,5,gene_description)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,6,picr_ensembl_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,7,chok1gs_ensembl_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,8,mRNA_ncbi_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,9,protein_ncbi_id)\n",
    "            time.sleep(5)\n",
    "            genes_sheet.update_cell(id,10,go_terms)\n",
    "        except:\n",
    "            print('Google API quota exceeded')\n",
    "            time.sleep(5)\n",
    "            continue\n",
    "            \n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30b59d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "handle = Entrez.efetch(db='gene', id='3979190', retmode='xml')\n",
    "record = Entrez.read(handle)[0]\n",
    "\n",
    "human_handle = Entrez.efetch(db='gene', id='3979190', retmode='xml')\n",
    "human_record = Entrez.read(human_handle)[0]\n",
    "if 'Entrezgene_comments' in human_record and 'Gene-commentary_comment' in human_record['Entrezgene_comments'][0]:\n",
    "    human_gene_description = human_record['Entrezgene_comments'][0]['Gene-commentary_comment'][0]['String']\n",
    "elif 'Entrezgene_summary' in human_record:\n",
    "    human_gene_description = human_record['Entrezgene_summary']\n",
    "else:\n",
    "    human_gene_description = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40d44cdb",
   "metadata": {},
   "source": [
    "#### Get the CHO genome and map the IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f04660af",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "import math\n",
    "import time\n",
    "import ssl\n",
    "from urllib.error import HTTPError\n",
    "import socket\n",
    "from Bio.Entrez.Parser import CorruptedXMLError\n",
    "import requests\n",
    "\n",
    "def get_uniprot_from_dict(dictionary, uniprotIDs, target_key, target_value1,  target_value2):\n",
    "    for key, value in dictionary.items():\n",
    "        if key == target_key and (value == target_value1 or value == target_value2):\n",
    "            uniprotIDs.append(dictionary.get('Dbtag_tag', {}).get('Object-id', {}).get('Object-id_str'))            \n",
    "        if isinstance(value, dict):\n",
    "            get_uniprot_from_dict(value, uniprotIDs, target_key, target_value1, target_value2)\n",
    "        elif isinstance(value, list):\n",
    "            for item in value:\n",
    "                if isinstance(item, dict):\n",
    "                    get_uniprot_from_dict(item, uniprotIDs, target_key, target_value1, target_value2)        \n",
    "    return uniprotIDs\n",
    "\n",
    "def get_human_mouse_orthologs(dictionary, orthologs, target_key, target_value1):\n",
    "    for key, value in dictionary.items():\n",
    "        if key == target_key and (value == target_value1):\n",
    "            orthologs.append(dictionary.get('Dbtag_tag', {}).get('Object-id', {}).get('Object-id_id'))        \n",
    "        if isinstance(value, dict):\n",
    "            get_human_mouse_orthologs(value, orthologs, target_key, target_value1)\n",
    "        elif isinstance(value, list):\n",
    "            for item in value:\n",
    "                if isinstance(item, dict):\n",
    "                    get_human_mouse_orthologs(item, orthologs, target_key, target_value1)        \n",
    "    return orthologs           \n",
    "\n",
    "\n",
    "u_g = pd.read_excel('../CHO_Genome.xlsx')\n",
    "GO_terms = pd.read_csv('orthologs&GO.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e34657fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_symbol_list = []\n",
    "\n",
    "gene_synoyms = []\n",
    "ncbi_protein_NP_weirdName = []\n",
    "ncbi_geneID_list = []\n",
    "ncbi_gene_exon_size = []\n",
    "# ncbi_genomic_location = []\n",
    "uniprot_IDs = []\n",
    "orthologs_genes = []\n",
    "orthologs_organism = []\n",
    "aa_sequences = []\n",
    "gene_record = []\n",
    "\n",
    "counter = 1\n",
    "for gene_EntrezID in u_g['NCBI GeneID'][counter:2]:\n",
    "# if True:\n",
    "#     gene_EntrezID = 100774298\n",
    "    protein_NP_weirdName = []\n",
    "    if not math.isnan(gene_EntrezID):   \n",
    "        gene_EntrezID = int(gene_EntrezID)                  \n",
    "        max_retries = 3\n",
    "        retries = 0\n",
    "        gene_found = True\n",
    "        while retries < max_retries and gene_found:\n",
    "\n",
    "            try:\n",
    "\n",
    "                search_results = Entrez.esearch(db=\"gene\", term = gene_EntrezID)            \n",
    "                record = Entrez.read(search_results)\n",
    "\n",
    "                # Check if any gene records were found\n",
    "                if int(record[\"Count\"]) > 0:\n",
    "\n",
    "                    # Get the Gene ID of the first result\n",
    "                    gene_id = record[\"IdList\"][0]\n",
    "\n",
    "                    # Fetch the gene information, maximum 3 tries\n",
    "                    fetch_tries = 0\n",
    "                    fetch_found = True\n",
    "                    while fetch_tries < 3 and fetch_found:\n",
    "\n",
    "                        gene_info = Entrez.efetch(db=\"gene\", id=gene_id, retmode=\"xml\")           \n",
    "                        fetch_tries += 1\n",
    "                        try:\n",
    "                            gene_record = Entrez.read(gene_info)\n",
    "                            fetch_found = False\n",
    "                        except CorruptedXMLError as xml_error:\n",
    "                            time.sleep(3)\n",
    "                            fetch_tries += 1\n",
    "\n",
    "                    # Get the gene name and gene NCBI ID\n",
    "                    if 'Gene-ref_locus' in gene_record[0]['Entrezgene_gene']['Gene-ref']:\n",
    "                        gene_symbol = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_locus']                        \n",
    "                    else:\n",
    "                        gene_symbol = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_locus-tag']\n",
    "                \n",
    "                    # Get the Gene Synonyms\n",
    "                    if 'Gene-ref_syn' in gene_record[0]['Entrezgene_gene']['Gene-ref']:\n",
    "                        gene_synonyms = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_syn']\n",
    "                    else:\n",
    "                        gene_synonyms = []\n",
    "                    gene_ncbiID = gene_record[0]['Entrezgene_track-info']['Gene-track']['Gene-track_geneid']\n",
    "\n",
    "                    # Get the Gene Description\n",
    "\n",
    "                    # Get UniprotIDs\n",
    "                    gene_uniprotIDs = (get_uniprot_from_dict(gene_record[0], [], 'Dbtag_db', 'UniProtKB/TrEMBL', 'UniProtKB/Swiss-Prot'))\n",
    "\n",
    "                    # Get the Gene Name\n",
    "                    gene_name = gene_record[0]['Entrezgene_gene']['Gene-ref']['Gene-ref_desc']\n",
    "\n",
    "                    # Get the chromosome\n",
    "                    chromosome = int(gene_record[0]['Entrezgene_source']['BioSource']['BioSource_subtype'][0]['SubSource_name'])\n",
    "\n",
    "                    # Get number of Exons\n",
    "                    exon_count = int(gene_record[0]['Entrezgene_properties'][0]['Gene-commentary_text'])\n",
    "\n",
    "                    # Get the Strand\n",
    "                    strand = gene_record[0]['Entrezgene_locus'][0]['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']['Seq-interval_strand']['Na-strand'].attributes['value']\n",
    "                    \n",
    "                    # Get Human and Mouse Orthologs\n",
    "                    orthologs_organism_list = []\n",
    "                    orthologs = []     \n",
    "                    if 'Entrezgene_comments' in gene_record[0]:\n",
    "                        for Entrezgene_comments_LOOP in gene_record[0]['Entrezgene_comments']:\n",
    "                            if 'Gene-commentary_comment' in Entrezgene_comments_LOOP:\n",
    "                                for Gene_commentary_comment_LOOP in Entrezgene_comments_LOOP['Gene-commentary_comment']:\n",
    "                                    if 'Gene-commentary_source' in Gene_commentary_comment_LOOP:\n",
    "                                        if 'Object-id_id' in Gene_commentary_comment_LOOP['Gene-commentary_source'][0]['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']:\n",
    "                                            orthologs.append(Gene_commentary_comment_LOOP['Gene-commentary_source'][0]['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_id'])\n",
    "                                        else:\n",
    "                                            orthologs.append(Gene_commentary_comment_LOOP['Gene-commentary_source'][0]['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str'])\n",
    "                                        orthologs_organism_list.append(Gene_commentary_comment_LOOP['Gene-commentary_source'][0]['Other-source_anchor'])                    \n",
    "                    human_ortholog_EntrezID = orthologs[orthologs_organism_list.index('human')]\n",
    "                    mouse_ortholog_EntrezID = orthologs[orthologs_organism_list.index('mouse')]\n",
    "                    \n",
    "                    # Get Gene Description\n",
    "                    search_results_human = Entrez.esearch(db=\"gene\", term=orthologs[orthologs_organism_list.index('human')])     \n",
    "                    record_human = Entrez.read(search_results_human)\n",
    "                    gene_id_human = record_human[\"IdList\"][0]\n",
    "                    gene_info_human = Entrez.efetch(db=\"gene\", id=gene_id_human, retmode=\"xml\")           \n",
    "                    gene_record_human = Entrez.read(gene_info_human)\n",
    "                    gene_description = gene_record_human[0]['Entrezgene_summary']\n",
    "\n",
    "                    # Get GO terms\n",
    "                    result = GO_terms[GO_terms['CHO GeneID'] == str(gene_EntrezID)]['GO_ids'].values\n",
    "                    if result.size > 0:\n",
    "                        go_terms = result[0]\n",
    "                    else:\n",
    "                        go_terms = None\n",
    "\n",
    "                    # Get the sequence                \n",
    "                    aa_bool = 'NOT Found AA_Seq'\n",
    "                    if protein_NP_weirdName != '-----':\n",
    "                        try:\n",
    "                            handle = Entrez.efetch(db=\"protein\", id=protein_NP_weirdName, rettype=\"fasta\")\n",
    "                            record = handle.read()\n",
    "                            amino_acid_sequence = ''.join(record.split('\\n')[1:])\n",
    "                            aa_bool = 'FOUND AA_Seq'\n",
    "                        except:\n",
    "                            handle = Entrez.efetch(db=\"nucleotide\", id=protein_NP_weirdName, rettype=\"fasta\")\n",
    "                            record = handle.read()\n",
    "                            amino_acid_sequence = ''.join(record.split('\\n')[1:])\n",
    "                            aa_bool = 'FOUND AA_Seq'\n",
    "                    # Get the COG terms\n",
    "                    cogs = []\n",
    "\n",
    "                # If there are no matches for the EntrezID at NCBI then return -----\n",
    "                else:\n",
    "                    gene_ncbiID = '-----'\n",
    "                    gene_symbol = '-----'\n",
    "                    gene_name_alsoknownas = '-----'\n",
    "                    protein_NP_weirdName = '-----'\n",
    "                    exon = '-----'\n",
    "                    # gene_location = '-----'                    \n",
    "                    uniprot_IDs = '-----'\n",
    "                    amino_acid_sequence = '-----'\n",
    "\n",
    "                gene_found = False\n",
    "                search_results.close()\n",
    "    \n",
    "            # Handle exceptions\n",
    "            except HTTPError as e:\n",
    "\n",
    "                retries += 1\n",
    "                time.sleep(3)\n",
    "            except ssl.SSLError as e:\n",
    "\n",
    "                retries += 1\n",
    "                time.sleep(3)\n",
    "            except ConnectionResetError as e:\n",
    "\n",
    "                retries += 1\n",
    "                time.sleep(3)\n",
    "            except socket.error as e:\n",
    "\n",
    "                retries += 1\n",
    "                time.sleep(3)\n",
    "    \n",
    "        # If the maximum fetch attempts reached, return ------\n",
    "        if retries >= max_retries:\n",
    "            gene_symbol = ncbi_gene\n",
    "            gene_synonyms = '-----'\n",
    "            gene_ncbiID = '-----'\n",
    "            protein_NP_weirdName = '-----'\n",
    "            exon = '-----'\n",
    "            # gene_location = '-----'\n",
    "            gene_uniprotIDs = '-----'\n",
    "            orthologs = '-----'\n",
    "            orthologs_organism_list = '-----'\n",
    "            amino_acid_sequence = '-----'\n",
    "    \n",
    "    # If for some reason the ncbi_gene is nto a number, then return the ncbi_gene for all cases\n",
    "    else:\n",
    "        gene_symbol = ncbi_gene\n",
    "        gene_synonyms = ncbi_gene\n",
    "        gene_ncbiID = ncbi_gene\n",
    "        protein_NP_weirdName = ncbi_gene\n",
    "        exon = ncbi_gene\n",
    "        # gene_location = ncbi_gene\n",
    "        amino_acid_sequence = ncbi_gene\n",
    "\n",
    "    # Append the lists with the data\n",
    "    # gene_symbol_list.append(gene_symbol)\n",
    "    # ncbi_geneName_alsoknownas_list.append(gene_name_alsoknownas)\n",
    "    # ncbi_geneID_list.append(gene_ncbiID)\n",
    "    # ncbi_protein_NP_weirdName.append(protein_NP_weirdName)\n",
    "    # ncbi_gene_exon_size.append(exon)\n",
    "    # # ncbi_genomic_location.append(gene_location)\n",
    "    # uniprot_IDs.append(gene_uniprotIDs)\n",
    "    # orthologs_genes.append(orthologs)\n",
    "    # orthologs_organism.append(orthologs_organism_list)\n",
    "    # aa_sequences.append(amino_acid_sequence)\n",
    "\n",
    "    # print(counter, gene_name, gene_name_alsoknownas, gene_ncbiID, exon, aa_bool, protein_NP_weirdName, gene_uniprotIDs, orthologs, orthologs_organism_list)\n",
    "    # print(gene_symbol, gene_synonyms, gene_EntrezID, gene_name, gene_description, go_terms, cogs, exon_count, chromosome, strand, human_ortholog_EntrezID, mouse_ortholog_EntrezID, assembly_specific_gene_symbol, gene_references)\n",
    "    \n",
    "    # Loop over the Genes\n",
    "    print(\"GENE: \", gene_symbol, \"-----\", gene_synonyms, \"-----\", gene_EntrezID, \"-----\", gene_name, \"-----\", gene_description[0:25], \"... -----\", go_terms[0:25], \"... -----\", cogs, \"-----\", exon_count, \"-----\", chromosome, \"-----\", strand, \"-----\", human_ortholog_EntrezID, \"-----\", mouse_ortholog_EntrezID)\n",
    "    \n",
    "    # Loop over the assembly specifics\n",
    "    for assembly_specific_gene in gene_record[0]['Entrezgene_locus']:\n",
    "        assembly_id = assembly_specific_gene['Gene-commentary_heading']\n",
    "        location_sequence = assembly_specific_gene['Gene-commentary_accession']\n",
    "        start_range = assembly_specific_gene['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']['Seq-interval_from']\n",
    "        end_range = assembly_specific_gene['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']['Seq-interval_to']\n",
    "        transcript_sequence_ID = assembly_specific_gene['Gene-commentary_products'][0]['Gene-commentary_accession']\n",
    "        print(\"ASSEMBLY SPECIFC GENE: S1, S2, \", gene_EntrezID, \"-----\", location_sequence, \"-----\", start_range, \"-----\", end_range, \"-----\", transcript_sequence_ID, \"-----\", assembly_id)\n",
    "\n",
    "        # Loop over the Transcripts and proteins\n",
    "        for transcripts in assembly_specific_gene['Gene-commentary_products']:\n",
    "            mRNA_handle = Entrez.efetch(db=\"nucleotide\", id=transcript_sequence_ID, rettype=\"fasta\")\n",
    "            mRNA_record = mRNA_handle.read()\n",
    "            mRNA_sequence = ''.join(mRNA_record.split('\\n')[1:])           \n",
    "            protein_IDs = []\n",
    "            for Entrez_Comments in gene_record[0]['Entrezgene_comments']:\n",
    "                if 'Gene-commentary_comment' in Entrez_Comments:\n",
    "                    if 'Gene-commentary_products' in Entrez_Comments['Gene-commentary_comment'][0]:\n",
    "                        for gene_products_protein in Entrez_Comments['Gene-commentary_comment'][0]['Gene-commentary_products']:\n",
    "                            uniprotID_list = []\n",
    "                            if 'Gene-commentary_products' in gene_products_protein['Gene-commentary_products'][0]:\n",
    "                                # gene_products_protein['Gene-commentary_products'][1]['Gene-commentary_products'][0]['Gene-commentary_source'][0]['Other-source_anchor']\n",
    "                                protein_sequence_ID =  gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_products'][0]['Gene-commentary_source'][0]['Other-source_anchor']\n",
    "                            else:\n",
    "                                protein_sequence_ID =  gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_accession']\n",
    "                            \n",
    "                            # protein_sequence_ID = gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_products'][0]['Gene-commentary_source'][0]['Other-source_anchor']\n",
    "                            # protein_sequence_ID = gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_source'][0]['Other-source_anchor']\n",
    "                            protein_handle = Entrez.efetch(db=\"protein\", id=protein_sequence_ID, rettype=\"fasta\")\n",
    "                            protein_record = protein_handle.read()\n",
    "                            protein_sequence = ''.join(protein_record.split('\\n')[1:])  \n",
    "\n",
    "                            if 'Gene-commentary_comment' in gene_products_protein['Gene-commentary_products'][0]:\n",
    "                                for gene_products_protein_othername in gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_comment']:        \n",
    "                                    if 'Gene-commentary_comment' in gene_products_protein_othername:\n",
    "                                        for gene_products_protein_uniprotID in gene_products_protein_othername['Gene-commentary_comment'][0]['Gene-commentary_source']:\n",
    "                                            if 'Other-source_src' in gene_products_protein_uniprotID:\n",
    "                                                if 'Object-id_str' in gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']:\n",
    "                                                    protein_uniprot_id = gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str']\n",
    "                                                    uniprotID_list.append(protein_uniprot_id)\n",
    "                            else:\n",
    "                                for gene_products_protein_uniprotID in gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_products'][0]['Gene-commentary_comment'][1]['Gene-commentary_comment'][0]['Gene-commentary_source']:        \n",
    "                                    protein_uniprot_id = gene_products_protein_uniprotID['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str']\n",
    "                                    # protein_uniprot_id = gene_products_protein['Gene-commentary_products'][0]['Gene-commentary_products'][0]['Gene-commentary_comment'][1]['Gene-commentary_comment'][0]['Gene-commentary_source'][1]['Other-source_src']['Dbtag']['Dbtag_tag']['Object-id']['Object-id_str']\n",
    "                                    \n",
    "                                    # print(protein_uniprot_id)\n",
    "                                    uniprotID_list.append(protein_uniprot_id)\n",
    "                                \n",
    "                            print(\"TRANSCRIPTS: \", transcript_sequence_ID, \"-----\", protein_sequence_ID, \"-----\", mRNA_sequence[0:25], '-----', assembly_id)\n",
    "                            # Loop over the UniprotIDs and get information\n",
    "                            subcellular_locations_list = []\n",
    "                            for UniprotID in uniprotID_list:\n",
    "                                name, weight, length, annotation_score, existance, subcellular_locations_list, tissue_specificity = UniProt_Data_Request(UniprotID)\n",
    "                                print(\"PROTEINS: \", UniprotID, \"-----\", name, \"-----\", existance, \"-----\", annotation_score, \"-----\", length, \"-----\", protein_sequence[0:25], \"-----\", weight, \"kDa -----\", subcellular_locations_list, \"-----\", tissue_specificity, \"-----\", protein_sequence_ID)\n",
    "    print(\"-----------------------------------------------------------------------------------------\")\n",
    "    counter += 1\n",
    "    # if counter % 1000 == 0:\n",
    "    #     data = {\n",
    "    #         'NCBI_ID': ncbi_geneID_list,\n",
    "    #         'Gene_Name': ncbi_geneName_list,\n",
    "    #         'NP_Name': ncbi_protein_NP_weirdName,\n",
    "    #         'Uniprot_ID': uniprot_IDs,\n",
    "    #         'Ortholog_Genes': orthologs_genes,\n",
    "    #         'Ortholog_Organisms': orthologs_organism,\n",
    "    #         'Exon Number': ncbi_gene_exon_size,\n",
    "    #         'Sequence': aa_sequences\n",
    "    #     }\n",
    "    #     df = pd.DataFrame(data)\n",
    "    #     file_path = 'GeneIDs_' + str(counter) + '.xlsx'\n",
    "    #     df.to_excel(file_path, index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1975cf6b",
   "metadata": {},
   "source": [
    "#### Combine different DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61015c42",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "All_Genes = pd.DataFrame()\n",
    "\n",
    "for i in range(1000, 35000, 1000):\n",
    "    Genes_temp = pd.read_excel('GeneIDs_' + str(i) + '.xlsx')\n",
    "    All_Genes = pd.concat([All_Genes, Genes_temp], ignore_index=True)\n",
    "All_Genes.to_excel('All_Genes.xlsx', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
