{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a0d39903",
   "metadata": {},
   "source": [
    "# Recon 3D CHO GPRs\n",
    "## Description\n",
    "This notebook describes the process by which we can mine information from Recon3D to add it to our reconstruction in CHO cells. First we map all the CHO GPRs corresponding to the Recon3D Human GPRs and a dataset containing all the reactions from Recon3D with their corresponding CHO GPRs is generated (1), then a dataset containing reactions from Recon3D that are not present in our reconstruction, along with annotated CHO GPRs, is generated in order to add it to our reconstruction (2)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "69c24edf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "import cobra\n",
    "from cobra.io.mat import load_matlab_model\n",
    "\n",
    "import gspread"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "067d6824",
   "metadata": {},
   "source": [
    "### 1. Finding CHO orthologs for Human GPRs in Recon3D\n",
    "Using the dataset generated in the **Orthologs** notebook, we can map all the CHO genes in the recon 3D dataset GPR column and generate a new column called \"CHO GPR\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5cb25025",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_reaction</th>\n",
       "      <th>m_gene_reaction_rule</th>\n",
       "      <th>m_metabolites</th>\n",
       "      <th>m_subsystem</th>\n",
       "      <th>m_gene</th>\n",
       "      <th>seq_uniprot</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>seq_file</th>\n",
       "      <th>seq_num_structures</th>\n",
       "      <th>seq_num_experimental_structures</th>\n",
       "      <th>struct_id</th>\n",
       "      <th>struct_chain_id</th>\n",
       "      <th>struct_chain_seq_coverage</th>\n",
       "      <th>struct_is_experimental</th>\n",
       "      <th>struct_pdb</th>\n",
       "      <th>struct_resolution</th>\n",
       "      <th>struct_chemicals</th>\n",
       "      <th>struct_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11DOCRTSLte</td>\n",
       "      <td>5243.1</td>\n",
       "      <td>h2o[c];atp[c];h[c];11docrtsl[c];11docrtsl[e];p...</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>5243.1</td>\n",
       "      <td>P08183-1</td>\n",
       "      <td>1280.0</td>\n",
       "      <td>P08183-1.fasta</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>3g61-A</td>\n",
       "      <td>A</td>\n",
       "      <td>80.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3g61</td>\n",
       "      <td>4.35</td>\n",
       "      <td>2J8</td>\n",
       "      <td>3g61-A_clean.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11DOCRTSTRNte</td>\n",
       "      <td>5243.1</td>\n",
       "      <td>atp[c];pi[c];11docrtstrn[c];h[c];11docrtstrn[e...</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>5243.1</td>\n",
       "      <td>P08183-1</td>\n",
       "      <td>1280.0</td>\n",
       "      <td>P08183-1.fasta</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>3g61-A</td>\n",
       "      <td>A</td>\n",
       "      <td>80.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3g61</td>\n",
       "      <td>4.35</td>\n",
       "      <td>2J8</td>\n",
       "      <td>3g61-A_clean.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12DHCHOLabc</td>\n",
       "      <td>8714.3 or 8647.1 or 8714.2 or 8714.1 or 1244.1...</td>\n",
       "      <td>atp[c];12dhchol[e];12dhchol[c];pi[c];h[c];h2o[...</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>1244.1</td>\n",
       "      <td>Q92887-1</td>\n",
       "      <td>1545.0</td>\n",
       "      <td>Q92887-1.fasta</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12DHCHOLabc</td>\n",
       "      <td>8714.3 or 8647.1 or 8714.2 or 8714.1 or 1244.1...</td>\n",
       "      <td>atp[c];12dhchol[e];12dhchol[c];pi[c];h[c];h2o[...</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>8647.1</td>\n",
       "      <td>O95342-1</td>\n",
       "      <td>1321.0</td>\n",
       "      <td>O95342-1.fasta</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12DHCHOLabc</td>\n",
       "      <td>8714.3 or 8647.1 or 8714.2 or 8714.1 or 1244.1...</td>\n",
       "      <td>atp[c];12dhchol[e];12dhchol[c];pi[c];h[c];h2o[...</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>10257.1</td>\n",
       "      <td>O15439-1</td>\n",
       "      <td>1325.0</td>\n",
       "      <td>O15439-1.fasta</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22107</th>\n",
       "      <td>r2525</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>gln_L[c];gln_L[e]</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>O75387-1</td>\n",
       "      <td>559.0</td>\n",
       "      <td>O75387-1.fasta</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>H16666-X</td>\n",
       "      <td>X</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_003618.1_model1_clean-X_clean.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22108</th>\n",
       "      <td>r2526</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>ser_L[e];ser_L[c]</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>O75387-1</td>\n",
       "      <td>559.0</td>\n",
       "      <td>O75387-1.fasta</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>H16666-X</td>\n",
       "      <td>X</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_003618.1_model1_clean-X_clean.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22109</th>\n",
       "      <td>r2532</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>asn_L[c];asn_L[e]</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>O75387-1</td>\n",
       "      <td>559.0</td>\n",
       "      <td>O75387-1.fasta</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>H16666-X</td>\n",
       "      <td>X</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_003618.1_model1_clean-X_clean.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22110</th>\n",
       "      <td>r2534</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>thr_L[c];thr_L[e]</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>O75387-1</td>\n",
       "      <td>559.0</td>\n",
       "      <td>O75387-1.fasta</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>H16666-X</td>\n",
       "      <td>X</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_003618.1_model1_clean-X_clean.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22111</th>\n",
       "      <td>r2535</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>hom_L[e];hom_L[c]</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>8501.1</td>\n",
       "      <td>O75387-1</td>\n",
       "      <td>559.0</td>\n",
       "      <td>O75387-1.fasta</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>H16666-X</td>\n",
       "      <td>X</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_003618.1_model1_clean-X_clean.pdb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22112 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          m_reaction                               m_gene_reaction_rule  \\\n",
       "0        11DOCRTSLte                                             5243.1   \n",
       "1      11DOCRTSTRNte                                             5243.1   \n",
       "2        12DHCHOLabc  8714.3 or 8647.1 or 8714.2 or 8714.1 or 1244.1...   \n",
       "3        12DHCHOLabc  8714.3 or 8647.1 or 8714.2 or 8714.1 or 1244.1...   \n",
       "4        12DHCHOLabc  8714.3 or 8647.1 or 8714.2 or 8714.1 or 1244.1...   \n",
       "...              ...                                                ...   \n",
       "22107          r2525                                             8501.1   \n",
       "22108          r2526                                             8501.1   \n",
       "22109          r2532                                             8501.1   \n",
       "22110          r2534                                             8501.1   \n",
       "22111          r2535                                             8501.1   \n",
       "\n",
       "                                           m_metabolites  \\\n",
       "0      h2o[c];atp[c];h[c];11docrtsl[c];11docrtsl[e];p...   \n",
       "1      atp[c];pi[c];11docrtstrn[c];h[c];11docrtstrn[e...   \n",
       "2      atp[c];12dhchol[e];12dhchol[c];pi[c];h[c];h2o[...   \n",
       "3      atp[c];12dhchol[e];12dhchol[c];pi[c];h[c];h2o[...   \n",
       "4      atp[c];12dhchol[e];12dhchol[c];pi[c];h[c];h2o[...   \n",
       "...                                                  ...   \n",
       "22107                                  gln_L[c];gln_L[e]   \n",
       "22108                                  ser_L[e];ser_L[c]   \n",
       "22109                                  asn_L[c];asn_L[e]   \n",
       "22110                                  thr_L[c];thr_L[e]   \n",
       "22111                                  hom_L[e];hom_L[c]   \n",
       "\n",
       "                    m_subsystem   m_gene seq_uniprot  seq_len        seq_file  \\\n",
       "0      Transport, extracellular   5243.1    P08183-1   1280.0  P08183-1.fasta   \n",
       "1      Transport, extracellular   5243.1    P08183-1   1280.0  P08183-1.fasta   \n",
       "2      Transport, extracellular   1244.1    Q92887-1   1545.0  Q92887-1.fasta   \n",
       "3      Transport, extracellular   8647.1    O95342-1   1321.0  O95342-1.fasta   \n",
       "4      Transport, extracellular  10257.1    O15439-1   1325.0  O15439-1.fasta   \n",
       "...                         ...      ...         ...      ...             ...   \n",
       "22107  Transport, extracellular   8501.1    O75387-1    559.0  O75387-1.fasta   \n",
       "22108  Transport, extracellular   8501.1    O75387-1    559.0  O75387-1.fasta   \n",
       "22109  Transport, extracellular   8501.1    O75387-1    559.0  O75387-1.fasta   \n",
       "22110  Transport, extracellular   8501.1    O75387-1    559.0  O75387-1.fasta   \n",
       "22111  Transport, extracellular   8501.1    O75387-1    559.0  O75387-1.fasta   \n",
       "\n",
       "       seq_num_structures  seq_num_experimental_structures struct_id  \\\n",
       "0                    21.0                             21.0    3g61-A   \n",
       "1                    21.0                             21.0    3g61-A   \n",
       "2                     0.0                              0.0       NaN   \n",
       "3                     0.0                              0.0       NaN   \n",
       "4                     0.0                              0.0       NaN   \n",
       "...                   ...                              ...       ...   \n",
       "22107                 2.0                              0.0  H16666-X   \n",
       "22108                 2.0                              0.0  H16666-X   \n",
       "22109                 2.0                              0.0  H16666-X   \n",
       "22110                 2.0                              0.0  H16666-X   \n",
       "22111                 2.0                              0.0  H16666-X   \n",
       "\n",
       "      struct_chain_id  struct_chain_seq_coverage  struct_is_experimental  \\\n",
       "0                   A                       80.5                     1.0   \n",
       "1                   A                       80.5                     1.0   \n",
       "2                 NaN                        NaN                     NaN   \n",
       "3                 NaN                        NaN                     NaN   \n",
       "4                 NaN                        NaN                     NaN   \n",
       "...               ...                        ...                     ...   \n",
       "22107               X                      100.0                     0.0   \n",
       "22108               X                      100.0                     0.0   \n",
       "22109               X                      100.0                     0.0   \n",
       "22110               X                      100.0                     0.0   \n",
       "22111               X                      100.0                     0.0   \n",
       "\n",
       "      struct_pdb  struct_resolution struct_chemicals  \\\n",
       "0           3g61               4.35              2J8   \n",
       "1           3g61               4.35              2J8   \n",
       "2            NaN                NaN              NaN   \n",
       "3            NaN                NaN              NaN   \n",
       "4            NaN                NaN              NaN   \n",
       "...          ...                ...              ...   \n",
       "22107        NaN                NaN              NaN   \n",
       "22108        NaN                NaN              NaN   \n",
       "22109        NaN                NaN              NaN   \n",
       "22110        NaN                NaN              NaN   \n",
       "22111        NaN                NaN              NaN   \n",
       "\n",
       "                                struct_file  \n",
       "0                          3g61-A_clean.pdb  \n",
       "1                          3g61-A_clean.pdb  \n",
       "2                                       NaN  \n",
       "3                                       NaN  \n",
       "4                                       NaN  \n",
       "...                                     ...  \n",
       "22107  NP_003618.1_model1_clean-X_clean.pdb  \n",
       "22108  NP_003618.1_model1_clean-X_clean.pdb  \n",
       "22109  NP_003618.1_model1_clean-X_clean.pdb  \n",
       "22110  NP_003618.1_model1_clean-X_clean.pdb  \n",
       "22111  NP_003618.1_model1_clean-X_clean.pdb  \n",
       "\n",
       "[22112 rows x 18 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Generate recon3d df from the recon3d supplementary data\n",
    "df1 = pd.read_excel('../../Data/GPR_Curation/recon3d_gprs.xlsx')\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dd88f21c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2026-03-06\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No defined compartments in model Recon3D. Compartments will be deduced heuristically using regular expressions.\n",
      "Using regular expression found the following compartments:c, e, g, i, l, m, n, r, x\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>Recon3D</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>30c4a50f0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>8399</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>13543</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>3697</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>111</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*biomass_reaction - 1.0*biomass_reaction_reverse_32a6c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>c, l, m, r, e, x, n, g, i</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model Recon3D at 0x30c4a50f0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Generate another df from the recon3d model\n",
    "recon3d_model = load_matlab_model('../../Data/Reconciliation/models/Recon3D_301.mat')\n",
    "recon3d_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c75d663f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2a646e81fb004a95b245b417fa769ccc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/13543 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/ipykernel_4135/3167771947.py:8: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  df2['m_reaction'] = df2['m_reaction'].str.replace(\"[\", \"_\")\n",
      "/var/folders/_x/tfg8s2ks4n1ftkkwzp5sqjpc0000gn/T/ipykernel_4135/3167771947.py:9: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  df2['m_reaction'] = df2['m_reaction'].str.replace(\"]\", \"\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_reaction</th>\n",
       "      <th>Reaction Name</th>\n",
       "      <th>m_metabolites</th>\n",
       "      <th>m_gene_reaction_rule</th>\n",
       "      <th>m_subsystem</th>\n",
       "      <th>Lower bound</th>\n",
       "      <th>Upper bound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10FTHF5GLUtl</td>\n",
       "      <td>5-Glutamyl-10Fthf Transport, Lysosomal</td>\n",
       "      <td>10fthf5glu[c] --&gt; 10fthf5glu[l]</td>\n",
       "      <td></td>\n",
       "      <td>Transport, lysosomal</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10FTHF5GLUtm</td>\n",
       "      <td>5-Glutamyl-10Fthf Transport, Mitochondrial</td>\n",
       "      <td>10fthf5glu[m] --&gt; 10fthf5glu[c]</td>\n",
       "      <td></td>\n",
       "      <td>Transport, mitochondrial</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10FTHF6GLUtl</td>\n",
       "      <td>6-Glutamyl-10Fthf Transport, Lysosomal</td>\n",
       "      <td>10fthf6glu[c] --&gt; 10fthf6glu[l]</td>\n",
       "      <td></td>\n",
       "      <td>Transport, lysosomal</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10FTHF6GLUtm</td>\n",
       "      <td>6-Glutamyl-10Fthf Transport, Mitochondrial</td>\n",
       "      <td>10fthf6glu[m] --&gt; 10fthf6glu[c]</td>\n",
       "      <td></td>\n",
       "      <td>Transport, mitochondrial</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10FTHF7GLUtl</td>\n",
       "      <td>7-Glutamyl-10Fthf Transport, Lysosomal</td>\n",
       "      <td>10fthf7glu[c] --&gt; 10fthf7glu[l]</td>\n",
       "      <td></td>\n",
       "      <td>Transport, lysosomal</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13538</th>\n",
       "      <td>CYOR_u10mi</td>\n",
       "      <td>CYOR_u10mi</td>\n",
       "      <td>2.0 ficytC[m] + 2.0 h[m] + q10h2[m] --&gt; 2.0 fo...</td>\n",
       "      <td>(7384.1 and 7388.1 and 4519.1 and 29796.2 and ...</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13539</th>\n",
       "      <td>Htmi</td>\n",
       "      <td>Htmi</td>\n",
       "      <td>h[i] --&gt; h[m]</td>\n",
       "      <td>9016.1 or 7352.2 or 7352.1 or 7351.1 or 9016.2...</td>\n",
       "      <td>Transport, mitochondrial</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13540</th>\n",
       "      <td>NADH2_u10mi</td>\n",
       "      <td>NADH2_u10mi</td>\n",
       "      <td>5.0 h[m] + nadh[m] + q10[m] --&gt; 4.0 h[i] + nad...</td>\n",
       "      <td>(4715.1 and 4720.1 and 4719.1 and 4700.1 and 5...</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13541</th>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>4.0 focytC[m] + 7.92 h[m] + o2[m] --&gt; 4.0 ficy...</td>\n",
       "      <td>1351.1 and 1347.1 and 1329.1 and 1327.1 and 34...</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13542</th>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>4.0 focytC[m] + 8.0 h[m] + o2[m] --&gt; 4.0 ficyt...</td>\n",
       "      <td>(7384.1 and 7388.1 and 4519.1 and 10975.1 and ...</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13543 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         m_reaction                               Reaction Name  \\\n",
       "0      10FTHF5GLUtl      5-Glutamyl-10Fthf Transport, Lysosomal   \n",
       "1      10FTHF5GLUtm  5-Glutamyl-10Fthf Transport, Mitochondrial   \n",
       "2      10FTHF6GLUtl      6-Glutamyl-10Fthf Transport, Lysosomal   \n",
       "3      10FTHF6GLUtm  6-Glutamyl-10Fthf Transport, Mitochondrial   \n",
       "4      10FTHF7GLUtl      7-Glutamyl-10Fthf Transport, Lysosomal   \n",
       "...             ...                                         ...   \n",
       "13538    CYOR_u10mi                                  CYOR_u10mi   \n",
       "13539          Htmi                                        Htmi   \n",
       "13540   NADH2_u10mi                                 NADH2_u10mi   \n",
       "13541       CYOOm3i                                     CYOOm3i   \n",
       "13542       CYOOm2i                                     CYOOm2i   \n",
       "\n",
       "                                           m_metabolites  \\\n",
       "0                        10fthf5glu[c] --> 10fthf5glu[l]   \n",
       "1                        10fthf5glu[m] --> 10fthf5glu[c]   \n",
       "2                        10fthf6glu[c] --> 10fthf6glu[l]   \n",
       "3                        10fthf6glu[m] --> 10fthf6glu[c]   \n",
       "4                        10fthf7glu[c] --> 10fthf7glu[l]   \n",
       "...                                                  ...   \n",
       "13538  2.0 ficytC[m] + 2.0 h[m] + q10h2[m] --> 2.0 fo...   \n",
       "13539                                      h[i] --> h[m]   \n",
       "13540  5.0 h[m] + nadh[m] + q10[m] --> 4.0 h[i] + nad...   \n",
       "13541  4.0 focytC[m] + 7.92 h[m] + o2[m] --> 4.0 ficy...   \n",
       "13542  4.0 focytC[m] + 8.0 h[m] + o2[m] --> 4.0 ficyt...   \n",
       "\n",
       "                                    m_gene_reaction_rule  \\\n",
       "0                                                          \n",
       "1                                                          \n",
       "2                                                          \n",
       "3                                                          \n",
       "4                                                          \n",
       "...                                                  ...   \n",
       "13538  (7384.1 and 7388.1 and 4519.1 and 29796.2 and ...   \n",
       "13539  9016.1 or 7352.2 or 7352.1 or 7351.1 or 9016.2...   \n",
       "13540  (4715.1 and 4720.1 and 4719.1 and 4700.1 and 5...   \n",
       "13541  1351.1 and 1347.1 and 1329.1 and 1327.1 and 34...   \n",
       "13542  (7384.1 and 7388.1 and 4519.1 and 10975.1 and ...   \n",
       "\n",
       "                     m_subsystem  Lower bound  Upper bound  \n",
       "0           Transport, lysosomal          0.0       1000.0  \n",
       "1       Transport, mitochondrial          0.0       1000.0  \n",
       "2           Transport, lysosomal          0.0       1000.0  \n",
       "3       Transport, mitochondrial          0.0       1000.0  \n",
       "4           Transport, lysosomal          0.0       1000.0  \n",
       "...                          ...          ...          ...  \n",
       "13538  Oxidative phosphorylation          0.0       1000.0  \n",
       "13539   Transport, mitochondrial          0.0       1000.0  \n",
       "13540  Oxidative phosphorylation          0.0       1000.0  \n",
       "13541  Oxidative phosphorylation          0.0       1000.0  \n",
       "13542  Oxidative phosphorylation          0.0       1000.0  \n",
       "\n",
       "[13543 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Generate another df from the recon3d model\n",
    "attributes = []\n",
    "for reaction in tqdm(recon3d_model.reactions):\n",
    "    attributes.append([reaction.id, reaction.name, reaction.reaction, reaction.gpr, \n",
    "                       reaction.subsystem, reaction.lower_bound, reaction.upper_bound])\n",
    "\n",
    "df2 = pd.DataFrame(data=attributes, columns=['m_reaction', 'Reaction Name', 'm_metabolites', 'm_gene_reaction_rule', 'm_subsystem', 'Lower bound', 'Upper bound'])\n",
    "df2['m_reaction'] = df2['m_reaction'].str.replace(\"[\", \"_\")\n",
    "df2['m_reaction'] = df2['m_reaction'].str.replace(\"]\", \"\")\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d2f1559",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unify df1 and df2 into one dataset containing all the necesary information\n",
    "recon3d = pd.concat([df2, df1])\n",
    "recon3d = recon3d.groupby('m_reaction').first()\n",
    "recon3d = recon3d.reset_index(drop = False)\n",
    "recon3d['m_reaction'] = recon3d['m_reaction'].str.replace('_hs$', '', regex=True) # Erase the _hs at the end of the reaction IDs\n",
    "recon3d['m_gene_reaction_rule'] = recon3d['m_gene_reaction_rule'].replace(np.nan,'',regex=True)\n",
    "recon3d.to_excel('../../Data/Reconciliation/datasets/recon3D_all_reactions.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbab91d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate orthologs dict from the dataset generated in the Orthologs notebook\n",
    "orthologs = pd.read_excel('../../Data/GPR_Curation/orthologs.xlsx', dtype=str)\n",
    "orthologs.fillna('', inplace=True)\n",
    "orthologs_dict = orthologs.set_index('Human GeneID')['CHO GeneID'].to_dict()\n",
    "orthologs_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a06165a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract GPR info from Recon3D and swap gene IDs from human to CHO\n",
    "import re\n",
    "cho_gpr = []\n",
    "\n",
    "for row in recon3d['m_gene_reaction_rule']:\n",
    "    row = str(row)\n",
    "    if row != \"\":\n",
    "        gpr = re.findall('[\\d.]*\\d+', row)\n",
    "        new_gpr = row\n",
    "        for g in gpr:\n",
    "            human_g = g.split('.')[0]\n",
    "            try:\n",
    "                cho_g = orthologs_dict[human_g]\n",
    "                if cho_g == '':\n",
    "                    cho_g = f'h{human_g}'\n",
    "            except:\n",
    "                cho_g = f'h{human_g}'\n",
    "            new_gpr = new_gpr.replace(g, cho_g)\n",
    "    elif row == '':\n",
    "        new_gpr = ''\n",
    "        \n",
    "    cho_gpr.append(new_gpr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "191cbe4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cho_gpr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da8d0d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate Recon3D dataset with the addition of GPR with CHO genes\n",
    "recon3d['CHO GPR'] = cho_gpr\n",
    "recon3d.to_excel('../../Data/GPR_Curation/recon3D_chogprs.xlsx')\n",
    "recon3d"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6976fa7",
   "metadata": {},
   "source": [
    "The dataset generated here will then be used in the **Reactions** notebook **Part 6**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91e8e076",
   "metadata": {},
   "source": [
    "### 2. Adittion of new reactions from Recon 3D\n",
    "Recon3D reactions with mapped CHO GPRs are added to our reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a21463e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset of Recon 3D dataset with information on CHO GPRs\n",
    "recon3d_cho = recon3d[recon3d['CHO GPR'] != '']\n",
    "recon3d_cho = recon3d_cho.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0aef670",
   "metadata": {},
   "outputs": [],
   "source": [
    "recon3d_cho['m_metabolites'] = recon3d_cho['m_metabolites'].str.replace(\"[\", \"_\", regex=True)\n",
    "recon3d_cho['m_metabolites'] = recon3d_cho['m_metabolites'].str.replace(\"]\", \"\", regex=True)\n",
    "recon3d_cho"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7fab964",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Retreive information from the reactions in our reconstruction in Google Sheets\n",
    "\n",
    "# give service account details to gspread\n",
    "sa = gspread.service_account(filename='../credentials.json')\n",
    "\n",
    "# sa is a gspread client, which can be used for connecting to the sheets\n",
    "# by using the open method and the sheet name.\n",
    "cho_recon = sa.open('CHO Network Reconstruction')\n",
    "\n",
    "# we also need to specify the page name before getting the data. In this case we use the Rxns and Metabolites sheet.\n",
    "rxns_sheet = cho_recon.worksheet('Rxns')\n",
    "\n",
    "# We can extract the data using the get_all_records method and create pd DataFrames\n",
    "\n",
    "# Reactions IDs, names, formulas, GPRs\n",
    "rxns = pd.DataFrame(rxns_sheet.get_all_records())\n",
    "del rxns[rxns.columns[0]]\n",
    "rxns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f48e041f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of all the reactions in our reconstruction\n",
    "cho_rxns_list = list(rxns['Reaction'])\n",
    "\n",
    "# List of all the reactions with CHO GPRs in Recon3D \n",
    "recon3D_rxns_list = list(recon3d_cho['m_reaction'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c235fffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of Recon3D reactions that are in our reconstruction\n",
    "rxns_in_cho = []\n",
    "\n",
    "# List of Recon3D reactions that are NOT in our reconstruction\n",
    "rxns_not_in_cho = []\n",
    "\n",
    "for rxn in recon3D_rxns_list:\n",
    "    if rxn in cho_rxns_list:\n",
    "        rxns_in_cho.append(rxn)\n",
    "    elif rxn not in cho_rxns_list:\n",
    "        rxns_not_in_cho.append(rxn)\n",
    "              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa26390",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generation of a dataset containing information about Recon3D reactions that are not in our reconstruction\n",
    "subset_df = recon3d_cho[recon3d_cho['m_reaction'].isin(rxns_not_in_cho)]\n",
    "subset_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f64e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This dataset contains all the reactions added from Recon 3D\n",
    "\n",
    "rxns_recon3d_toadd = subset_df[['m_reaction', 'Reaction Name', 'm_metabolites', 'm_subsystem', 'CHO GPR', 'Lower bound', 'Upper bound']]\n",
    "rxns_recon3d_toadd.to_excel('../../Data/Reconciliation/datasets/rxns_recon3d_toadd.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1a09993",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
